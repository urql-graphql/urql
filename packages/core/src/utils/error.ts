import { GraphQLError } from 'graphql';

const generateErrorMessage = (
  networkErr?: Error,
  graphQlErrs?: GraphQLError[]
) => {
  let error = '';
  if (networkErr) {
    return (error = `[Network] ${networkErr.message}`);
  }

  if (graphQlErrs) {
    for (const err of graphQlErrs) {
      error += `[GraphQL] ${err.message}\n`;
    }
  }

  return error.trim();
};

const rehydrateGraphQlError = (error: any): GraphQLError => {
  if (typeof error === 'string') {
    return new GraphQLError(error);
  } else if (typeof error === 'object' && error.message) {
    return new GraphQLError(
      error.message,
      error.nodes,
      error.source,
      error.positions,
      error.path,
      error,
      error.extensions || {}
    );
  } else {
    return error as any;
  }
};

/** An error which can consist of GraphQL errors and Network errors. */
export class CombinedError extends Error {
  public name: string;
  public message: string;
  public graphQLErrors: GraphQLError[];
  public networkError?: Error;
  public response?: any;

  constructor({
    networkError,
    graphQLErrors,
    response,
  }: {
    networkError?: Error;
    graphQLErrors?: Array<string | Partial<GraphQLError> | Error>;
    response?: any;
  }) {
    const normalizedGraphQLErrors = (graphQLErrors || []).map(
      rehydrateGraphQlError
    );
    const message = generateErrorMessage(networkError, normalizedGraphQLErrors);

    super(message);

    this.name = 'CombinedError';
    this.message = message;
    this.graphQLErrors = normalizedGraphQLErrors;
    this.networkError = networkError;
    this.response = response;
  }

  toString() {
    return this.message;
  }
}
