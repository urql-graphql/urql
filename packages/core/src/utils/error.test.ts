import { describe, it, expect } from 'vitest';
import { CombinedError } from './error';

describe('CombinedError', () => {
  it('can be instantiated with graphQLErrors', () => {
    const err = new CombinedError({
      graphQLErrors: [],
    });

    expect(err.name).toBe('CombinedError');
  });

  it('behaves like a normal Error', () => {
    const err = new CombinedError({
      networkError: new Error('test'),
    });

    expect(err instanceof CombinedError).toBe(true);
    expect(err instanceof Error).toBe(true);
    expect(Error.prototype.isPrototypeOf(err)).toBe(true);
    expect('' + err).toMatchSnapshot();
  });

  it('accepts graphQLError messages and generates a single message from them', () => {
    const graphQLErrors = ['Error Message A', 'Error Message B'];

    const err = new CombinedError({ graphQLErrors });

    expect(err.message).toBe(
      `
[GraphQL] Error Message A
[GraphQL] Error Message B
    `.trim()
    );

    expect(err.graphQLErrors).toEqual(graphQLErrors.map(x => new Error(x)));
  });

  it('accepts a network error and generates a message from it', () => {
    const networkError = new Error('Network Shenanigans');
    const err = new CombinedError({ networkError });

    expect(err.message).toBe(`[Network] ${networkError.message}`);
  });

  it('accepts actual errors for graphQLError', () => {
    const graphQLErrors = [
      new Error('Error Message A'),
      new Error('Error Message B'),
    ];

    const err = new CombinedError({ graphQLErrors });

    expect(err.message).toBe(
      `
[GraphQL] Error Message A
[GraphQL] Error Message B
    `.trim()
    );

    expect(err.graphQLErrors).toEqual(graphQLErrors);
  });

  it('passes graphQLErrors through as a last resort', () => {
    const graphQLErrors = [{ x: 'y' }] as any;
    const err = new CombinedError({ graphQLErrors });

    expect(err.graphQLErrors).toEqual(graphQLErrors);
  });

  it('accepts a response that is attached to the resulting error', () => {
    const response = {};
    const err = new CombinedError({
      graphQLErrors: [],
      response,
    });

    expect(err.response).toBe(response);
  });
});
