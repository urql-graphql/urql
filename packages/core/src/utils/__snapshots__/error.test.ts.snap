// Vitest Snapshot v1

exports[`CombinedError > behaves like a normal Error 1`] = `"[Network] test"`;
